### ISSUE TITLE
`kl_weight` specification in probabilistic layer regression example

### ISSUE THREAD

**Issue by kevinykuo** [tensorflow/probability#396]
Why are we dividing by the number of samples and not the batch size at https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Probabilistic_Layers_Regression.ipynb?short_path=2d2ce85#L770 ?

```
# Build model.
model = tf.keras.Sequential([
  tfp.layers.DenseVariational(1, posterior_mean_field, prior_trainable, kl_weight=1/x.shape[0]),
  tfp.layers.DistributionLambda(lambda t: tfd.Normal(loc=t, scale=1)),
])

# Do inference.
model.compile(optimizer=tf.optimizers.Adam(learning_rate=0.01), loss=negloglik)
model.fit(x, y, epochs=1000, verbose=False);

# Profit.
[print(np.squeeze(w.numpy())) for w in model.weights];
yhat = model(x_tst)
assert isinstance(yhat, tfd.Distribution)
```

**Comment by Himscipy** [comment]
Hi, This is because you are passing the complete data-set at one go, I don't see any batch iterations done in this example case you are pointing too.  You are right that KL terms should be scaled according to the batch size.

**Comment by kevinykuo** [comment]
`fit()` defaults to batch size of 32, no?

**Comment by Himscipy** [comment]
I don't know if fit has a default batch size.  But in that case as well the x.shape[0] will ensure that the kl is divided by the number of training samples.

**Comment by kevinykuo** [comment]
OK then it seems like we should either grab the batch size at runtime or make sure the batch size divides the number of samples. @jvdillon could you confirm?

**Comment by jvdillon** [comment]
Yep. Note that Probabilistic_Layers_Regression.ipynb was updated to use
`kl_weight=1/x.shape[0]` which is ` kl_weight=1/num_training_points` which
I think is the same idea.

*From: *Kevin Kuo <notifications@github.com>
*Date: *Tue, May 7, 2019 at 3:58 PM
*To: *tensorflow/probability
*Cc: *Joshua V. Dillon, Mention

OK then it seems like we should either grab the batch size at runtime or

**Comment by kevinykuo** [comment]
Wouldn't `1/num_training_points` be incorrect in general though since batch_size can be much smaller than num_training_points? In the example we're talking about 1/150 vs 1/32.

**Comment by skeydan** [comment]
If I modify the notebook to include a callback that prints the loss

```
c = tf.keras.callbacks.LambdaCallback(
    on_batch_end=lambda batch,logs: print(str(batch) + ": " + str(logs["loss"])))

model = tf.keras.Sequential([
  tfp.layers.DenseVariational(1, posterior_mean_field, prior_trainable, kl_weight=1/x.shape[0]),
  tfp.layers.DistributionLambda(lambda t: tfd.Normal(loc=t, scale=1)),
])

# Do inference.
model.compile(optimizer=tf.optimizers.Adam(learning_rate=0.01), loss=negloglik)
model.fit(x, y, epochs=1, verbose=False, callbacks=[c]);

```

and fit it for one epoch, I see

```
0: 2462.2026
1: 30.330355
2: 1666.2606
3: 33.58991
4: 13.362394
```

But the KL loss gets scaled in every `call`:

```
  def call(self, inputs):
    dtype = tf.as_dtype(self.dtype or tf.keras.backend.floatx())
    inputs = tf.cast(inputs, dtype, name='inputs')

    q = self._posterior(inputs)
    r = self._prior(inputs)
    self.add_loss(self._kl_divergence_fn(q, r))

```

So for every batch, of (default) size `32`, we divide by the size of the overall dataset (`150`).

It seems like we really want to divide by the batch size instead.

**Comment by junpenglao** [comment]
I think dividing by the size of the overall dataset is correct. The loss in each mini-batch is the sum of k (number of observation in the batch) data point, and you would want to scale it proportion to n (total number of observation).

**Comment by skeydan** [comment]
Oh, that makes sense! Thanks for the explanation!

**Comment by daniloefl** [comment]
Dear @kevinykuo and @junpenglao ,

I was also wondering about this, and although I agree that the KL divergence should be scaled by the number of samples, what is important is that the output loss (`-dist.log_prob(y_true)`) is such that the value minimized is still proportional to - ELBO.

While the KL divergence is summed for each batch sample and for each batch (and then scaled down appropriately using the `kl_weight`), the negative log-probability needs to be equivalently summed over all batches and all samples (`-ELBO = -sum log-prob + KL divergence`).

This is not explicitly done and looking at the code [1], the default behaviour is to average over the batch. As a result, the total loss for an epoch would be `- 1/Nb * sum_N log-prob + 1/N sum_N KL divergence`. 

Even if the total loss is averaged over the batches, the resulting total loss is not proportional to the ELBO (the reduction over batches only changes a multiplicative constant in the total loss, as long as it is done for both terms in the same way).

I actually hope I am wrong, but if I am not, one needs only to add `tf.reduce_sum(nll)` in the loss function definition. I have tried adding `tf.reduce_mean(nll)` in it and the loss function does not changes significantly, as I expect from the above, but it does change if I use `tf.reduce_sum(nll)` (see [2]).

In any case, please let me know if I have misunderstood something.

Best regards,
Danilo

[1] https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/engine/training.py#L1741

Note that `SUM_OVER_BATCH_SIZE` is the default and it corresponds to an average over the batch:
https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/utils/losses_utils.py#L58

[2] Starting loss function with the current notebook setup (no clear definition in the loss function over what to do with the batch axis):
```
Epoch 1/10
150/150 [==============================] - 0s 1ms/sample - loss: 562.2356
Epoch 2/10
150/150 [==============================] - 0s 68us/sample - loss: 458.5881
Epoch 3/10
150/150 [==============================] - 0s 65us/sample - loss: 358.3324
Epoch 4/10
150/150 [==============================] - 0s 94us/sample - loss: 440.9295
Epoch 5/10
150/150 [==============================] - 0s 71us/sample - loss: 663.6378
Epoch 6/10
150/150 [==============================] - 0s 69us/sample - loss: 254.1046
Epoch 7/10
150/150 [==============================] - 0s 64us/sample - loss: 475.4317
Epoch 8/10
150/150 [==============================] - 0s 61us/sample - loss: 1313.9629
Epoch 9/10
150/150 [==============================] - 0s 59us/sample - loss: 818.2486
Epoch 10/10
150/150 [==============================] - 0s 69us/sample - loss: 999.9185
```

Using `tf.reduce_mean()` here `negloglik = lambda y, rv_y: -tf.reduce_mean(rv_y.log_prob(y))`:

```
Epoch 1/10
150/150 [==============================] - 0s 1ms/sample - loss: 240.1954
Epoch 2/10
150/150 [==============================] - 0s 65us/sample - loss: 135.6446
Epoch 3/10
150/150 [==============================] - 0s 68us/sample - loss: 1946.0204
Epoch 4/10
150/150 [==============================] - 0s 69us/sample - loss: 2539.7345
Epoch 5/10
150/150 [==============================] - 0s 81us/sample - loss: 431.5072
Epoch 6/10
150/150 [==============================] - 0s 67us/sample - loss: 1237.3390
Epoch 7/10
150/150 [==============================] - 0s 64us/sample - loss: 259.5903
Epoch 8/10
150/150 [==============================] - 0s 62us/sample - loss: 1208.8005
Epoch 9/10
150/150 [==============================] - 0s 63us/sample - loss: 1321.6166
Epoch 10/10
150/150 [==============================] - 0s 63us/sample - loss: 616.1644
```

Using `tf.reduce_sum()` here `negloglik = lambda y, rv_y: -tf.reduce_sum(rv_y.log_prob(y))`:

```
Epoch 1/10
150/150 [==============================] - 0s 1ms/sample - loss: 17829.9767
Epoch 2/10
150/150 [==============================] - 0s 70us/sample - loss: 52574.5798
Epoch 3/10
150/150 [==============================] - 0s 85us/sample - loss: 51067.9562
Epoch 4/10
150/150 [==============================] - 0s 68us/sample - loss: 14182.6555
Epoch 5/10
150/150 [==============================] - 0s 77us/sample - loss: 32011.6627
Epoch 6/10
150/150 [==============================] - 0s 75us/sample - loss: 16739.6968
Epoch 7/10
150/150 [==============================] - 0s 79us/sample - loss: 58137.4190
Epoch 8/10
150/150 [==============================] - 0s 85us/sample - loss: 37134.6132
Epoch 9/10
150/150 [==============================] - 0s 80us/sample - loss: 26681.6911
Epoch 10/10
150/150 [==============================] - 0s 80us/sample - loss: 7566.2281
```
